### PR TITLE
Support typescript

### PR DESCRIPTION
**Pull request by alexduf**
## What does this change?

This extends the build to also generate and publish typescript files from the thrift definition.
It using the new sbt-plugin and scrooge extension I've implemented [here](https://github.com/guardian/scrooge-extras)

## How can we measure success?

Developers are able to pull an npm package that contains the CAPI models

I'm still unsure how smooth the release will be as it's impossible to test without actually releasing :)

### CODE REVIEW

**Comment by jonathonherbert** [comment]
@alexduf re: test releases, could you use a [prerelease version](https://docs.npmjs.com/cli/version) to mark a release as not yet ready for consumption?

**Comment by alexduf** [comment]
@jonathonherbert I guess I could do that manually rather than using the sbt-plugin at first, and then re-release if we're happy?

If I understand prerealeases correctly, it's a matter of suffixing the version number with a string, like `1.0.0-SNAPSHOT`? If so when releasing we can type whatever version number we want

**Comment by jonathonherbert** [comment]
@alexduf I think this is very sensible – there's also `npm link` to test consumption of a package locally before release. Super happy to help test if needed 👍

**Comment by alexduf** [comment]
I've tested on the content-atom repository by publishing libraries locally and importing them.

It compiles so it looks all good!

For instance the review atom successfully generates:
```typescript
import {
    Film,
    FilmSerde,
} from '@guardian/content-entity-model/film/film'
```

**Review by jonathonherbert — APPROVED** [review]
@alexduf I've tested this locally and the type definitions are easily accessible 👍 

I'm approving this on that basis to unblock, but I'm aware that there's serialisation/deserialisation work in here, too, that I haven't yet tested – and that this repo is primarily used by other teams! @JustinPinner @davidfurey (e.g.) do you have anything to add?

File: build.sbt
Comment (JustinPinner) [inline build.sbt]: Just checking that publishing is disabled intentionally here for production merge?

**Comment by JustinPinner** [comment]
How did you get around the naming/reserved word issues you mentioned in the last meeting? I think we discussed "French thrift" as a solution, and I'm probably missing something but I don't see that mentioned here?

File: build.sbt
Comment (alexduf) [reply build.sbt]: Good catch that was there for a test. I've removed it

**Comment by alexduf** [comment]
@JustinPinner I've actually [re-implented the code generation](https://github.com/guardian/scrooge-extras). So we don't rely on `thrift` or its french fork :)

There's an explanation here if you want: https://github.com/guardian/scrooge-extras/pull/3

**Comment by JustinPinner** [comment]
Ok, I'm not going to pretend that I'm familiar with everything here but from what I can see, it's an additional step in the release process that can be considered as an isolated, and therefore low-risk, alteration. I'm assuming that the namespaces added to the few thrift files that have changed have no effect beyond the scope of the new features and everything that was in place before remains unaffected. Further, clients will have to be updated to consume the new published libraries so releasing this now won't have any effect at all on currently functioning systems. If that is all correct then I have no objections to merging this.

**Review by JustinPinner — APPROVED** [review]
On the basis that my assumptions (see other comment) are accurate: 👍

**Comment by JustinPinner** [comment]
@Mullefa I know you're keen to understand this too, so let's hold back merging until you're able to approve it too?

**Comment by alexduf** [comment]
Yes:
 - no breaking change for the existing scala consumers, new typescript clients will have to actively pull the new dependency
 - the namespaces are indeed only affecting the typescript generation
 - separate release process. I actually tried to integrate it with sbt-release and failed, but I don't think that's a blocker

**Comment by Mullefa** [comment]
hey - @JustinPinner thanks for the note, but if it's got multiple approvals already, don't let me hold this PR back from shipping 🚢 

(though I'll have a look anyway - looks cool 👏 )

**Review by Mullefa — APPROVED** [review]

**Pull request merged**
